packages/history/index.ts

```diff
@@ -1,25 +1,94 @@
-const PopAction = 'POP';
-const PushAction = 'PUSH';
-const ReplaceAction = 'REPLACE';
+////////////////////////////////////////////////////////////////////////////////
+// TYPES
+////////////////////////////////////////////////////////////////////////////////
+
+export type Action = 'POP' | 'PUSH' | 'REPLACE';
+
+export type Path = string;
+export type PathPieces = {
+  pathname?: string;
+  search?: string;
+  hash?: string;
+};
+
+export type State = object;
+export interface Location<S extends State = State> extends PathPieces {
+  pathname: string;
+  search: string;
+  hash: string;
+  state?: S;
+  key?: string;
+}
+
+export interface Update<S extends State = State> {
+  action: Action;
+  location: Location<S>;
+}
+export interface Listener<S extends State = State> {
+  (update: Update<S>): void;
+}
+export type Unlistener = () => void;
+
+export interface Transaction<S extends State = State> extends Update {
+  retry(): void;
+}
+export interface Blocker<S extends State = State> {
+  (tx: Transaction<S>): void;
+}
+export type Unblocker = () => void;
+
+export interface History<S extends State = State> {
+  action: Action;
+  location: Location<S>;
+  createHref(to: Path | PathPieces): string;
+  push(to: Path | PathPieces, state?: State): void;
+  replace(to: Path | PathPieces, state?: State): void;
+  go(n: number): void;
+  back(): void;
+  forward(): void;
+  listen(listener: Listener<S>): Unlistener;
+  block(blocker: Blocker<S>): Unblocker;
+}
+export interface MemoryHistory<S extends State = State> extends History<S> {
+  index: number;
+}
+
+type HistoryState = {
+  usr?: State;
+  key?: string;
+  idx?: number;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+// CONSTANTS
+////////////////////////////////////////////////////////////////////////////////
+
+const PopAction: Action = 'POP';
+const PushAction: Action = 'PUSH';
+const ReplaceAction: Action = 'REPLACE';
 
 const BeforeUnloadEventType = 'beforeunload';
 const HashChangeEventType = 'hashchange';
 const PopStateEventType = 'popstate';
 
-const readOnly = __DEV__ ? obj => Object.freeze(obj) : obj => obj;
+const readOnly = __DEV__ ? (obj: any) => Object.freeze(obj) : (obj: any) => obj;
```

Comment:
Perhaps this should be generic:

```ts
const readOnly = __DEV__ ? <T>(obj: T) => Object.freeze(obj) : <T>(obj: T) => obj;
```

or

```ts
const readOnly = <T>(obj: T) => __DEV__ ? Object.freeze(obj) : obj;
```

`any` completely disables type checking so it should be avoided, e.g.

```ts
declare const myObject: any;

myObject.foo.bar.baz // no compile time error, but runtime exception!!
```

For this reason, `unknown` or generics should be used instead (depending on the use case)

Reply:
I like the first one! 🕺

Reply:
```ts
const readOnly: typeof Object.freeze = __DEV__ ? Object.freeze : ((obj:any) => obj);
```

I think this one is better. This means:

```ts
const readOnly: {
    <T>(a: T[]): readonly T[];
    <T extends Function>(f: T): T;
    <T>(o: T): Readonly<T>;
}
```

Reply:
This is a great suggestion, thank you @OliverJAsh. I'll fix it.